src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DetectAmbiguousRoutes.cs

```diff
@@ -0,0 +1,215 @@
+    {
+        if (mapOperations.IsEmpty)
+        {
+            return;
+        }
+
+        var groupedByParent = mapOperations
+            .Select(kvp => new { MapOperation = kvp.Key, ResolvedOperation = ResolveOperation(kvp.Key.Operation, wellKnownTypes) })
+            .Where(u => u.ResolvedOperation != null && !u.MapOperation.RouteUsageModel.UsageContext.HttpMethods.IsDefault)
+            .GroupBy(u => new MapOperationGroupKey(u.MapOperation.Builder, u.ResolvedOperation!, u.MapOperation.RouteUsageModel.RoutePattern, u.MapOperation.RouteUsageModel.UsageContext.HttpMethods));
+
+        foreach (var ambigiousGroup in groupedByParent.Where(g => g.Count() >= 2))
+        {
+            foreach (var ambigiousMapOperation in ambigiousGroup)
+            {
+                var model = ambigiousMapOperation.MapOperation.RouteUsageModel;
+
+                context.ReportDiagnostic(Diagnostic.Create(
+                    DiagnosticDescriptors.AmbiguousRouteHandlerRoute,
+                    model.UsageContext.RouteToken.GetLocation(),
+                    model.RoutePattern.Root.ToString()));
+            }
+        }
+    }
+
+    private static IOperation? ResolveOperation(IOperation operation, WellKnownTypes wellKnownTypes)
+    {
+        // We want to group routes in a block together because we know they're being used together.
+        // There are some circumstances where we still don't want to use the route, either because it is only conditionally
+        // being called, or the IEndpointConventionBuilder returned from the method is being used. We can't accurately
+        // detect what extra endpoint metadata is being added to the routes.
+        //
+        // Don't use route endpoint if:
+        // - It's in a conditional statement.
+        // - It's in a coalesce statement.
+        // - It's has methods called on it.
+        // - It's assigned to a variable.
+        // - It's an argument to a method call, unless in a known safe method.
+        var current = operation;
+        if (current.Parent is IArgumentOperation argumentOperation &&
+            argumentOperation.Parent is IInvocationOperation invocationOperation &&
+            IsAllowedEndpointBuilderMethod(invocationOperation, wellKnownTypes))
+        {
+            return ResolveOperation(invocationOperation, wellKnownTypes);
+        }
+
+        while (current != null)
+        {
+            if (current.Parent is IBlockOperation blockOperation)
+            {
+                return blockOperation;
+            }
+            else if (current.Parent is IConditionalOperation ||
+                current.Parent is ICoalesceOperation ||
+                current.Parent is IAssignmentOperation ||
+                current.Parent is IArgumentOperation ||
+                current.Parent is IInvocationOperation)
```

Comment:
Stylistic suggestion

```suggestion
            else if (current.Parent is IConditionalOperation or
                ICoalesceOperation or
                IAssignmentOperation or
                IArgumentOperation or
                IInvocationOperation)
```